### ISSUE TITLE
Suggestion for balanced graph cut

### ISSUE THREAD

**Issue by ljmartin** [sknetwork-team/scikit-network#259]
* scikit-network version: 0.12.1
* Python version: 3.7
* Operating System: ubuntu

Hi scikit-network,
Thanks for this library. It's really fast and well written and has been very useful for clustering large graphs (~10-100K nodes). The Paris cluster algorithm works really well for my purposes.

Instead of straight cuts, I've been using balanced cuts, which aim to keep all clusters a similar size. Essentially it uses the `sknetwork.hierarchy.straight_cut` multiple times while increasing the number of clusters until any new cluster gets below some maximum size. Heres an example:
https://gist.github.com/ljmartin/7533d3837bad597b0b2971488808d87b

If there's any interest then Ill submit a PR. At the moment I'm trying to get this (and `straight_cut`) to `jit` because it gets quite slow to call `straight_cut` multiple times on large graphs.

Inspiration came from https://github.com/vreyespue/ward_cut_tree_balanced which is now being merged into scipy. However doing a `scipy.cluster.hierarchy.cut_tree` on a 10,000 node graph gives me a memory error, hence the successive calls to your `straight_cut` method instead. 
cheers
lewis

**Comment by ljmartin** [comment]
OK, I think I've got a version of `straight_cut` that will be more amenable to `numba`fying, and at least to me it makes more intuitive sense but that may be numpy bias. 

However I haven't got it quite working. Would appreciate some help if anyone is a `numba` expert?

Code:
```
def do_cut(dendrogram, n_clusters=2):
    n_nodes = dendrogram.shape[0]+1
    labels = np.arange(n_nodes)
    last_cluster_id = max(labels)+1
    for i in range(n_nodes-n_clusters):
        indices = dendrogram[i][:2].astype(int)
        labels[np.isin(labels, indices)]=last_cluster_id
        last_cluster_id+=1
    labels = labels-labels.min()
    return labels
```


Self-contained example:
```
import numpy as np
import sknetwork as skn

adjacency = skn.data.karate_club()
paris = skn.hierarchy.Paris(engine='python')
dendrogram = paris.fit_transform(adjacency)
labels = skn.hierarchy.straight_cut(dendrogram)

def do_cut(dendrogram, n_clusters=2):
    n_nodes = dendrogram.shape[0]+1
    labels = np.arange(n_nodes)
    last_cluster_id = max(labels)+1
    for i in range(n_nodes-n_clusters):
        indices = dendrogram[i][:2].astype(int)
        labels[np.isin(labels, indices)]=last_cluster_id
        last_cluster_id+=1
    labels = labels-labels.min()
    return labels


do_cut(dendrogram) == skn.hierarchy.cuts.straight_cut(dendrogram)
```

output:
```
array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True])
```
cheers

**Comment by nathandelara** [comment]
Thank you for your comment! We will take a look at this first thing next week. 

What speed improvement did you get from numba in your setting ?

**Comment by ljmartin** [comment]
Good question! I tested it and found out that my version of straight_cut is far, far slower on large graphs. The dict-based version you already have is really fast. 

So I realised `straight_cut` can actually be slightly altered to give a balanced cut, and it gets the benefits of that speed.

So below is a better version of `balanced_cut`, based on your `straight_cut`, which can handle a graph of ~100000 nodes in 200ms:
```
def balanced_cut(dendrogram, max_cluster_size):
    n_nodes = dendrogram.shape[0] + 1
    labels = np.zeros(n_nodes, dtype=int)
    cluster = {node: [node] for node in range(n_nodes)}
    completed_clusters = list()
    
    for t in range(n_nodes - 1):
        currentID = n_nodes+t
        left = cluster[int(dendrogram[t][0])]
        right = cluster[int(dendrogram[t][1])]
        if len(left)+len(right) > max_cluster_size:
            for clust in [left, right]:
                if len(clust)<max_cluster_size:
                    completed_clusters.append(clust)
                    
        cluster[currentID] = cluster.pop(int(dendrogram[t][0])) + cluster.pop(int(dendrogram[t][1]))
    
    for count, indices in enumerate(completed_clusters):
        labels[indices]=count
    return labels
```

I did a little comparison to the `ward_cut_tree_balanced` from the scipy PRs showing that this version gets the same results but faster:
https://gist.github.com/ljmartin/922c13fcb2a59e11d3f98fb5690bab6c


It's would be compatible with having a single function just called `cut` that has `balanced` or `straight` options.

cheers

**Comment by tbonald** [comment]
Thanks for the suggestion.
Below is a proposal, close to your code:

`def balanced_cut(dendrogram: np.ndarray, max_cluster_size: int = 2, sorted_clusters: bool = True) -> np.ndarray:
    """
    Extract the clustering from the dendrogram with some maximum size for each cluster.

    Parameters
    ----------
    dendrogram:
        Dendrogram
    max_cluster_size :
        Maximum size of each cluster.
    sorted_clusters :
        If True, sort labels in decreasing order of cluster size.

    Returns
    -------
    labels : np.ndarray
        Cluster index of each node.
    """
    n_nodes = dendrogram.shape[0] + 1
    if max_cluster_size < 2 or max_cluster_size > n_nodes:
        raise ValueError("The maximum cluster size must be between 2 and the number of nodes.")

    labels = np.zeros(n_nodes, dtype=int)
    cluster = {node: [node] for node in range(n_nodes)}
    for t in range(n_nodes - 1):
        left = int(dendrogram[t][0])
        right = int(dendrogram[t][1])
        if left in cluster and right in cluster and len(cluster[left]) + len(cluster[right]) <= max_cluster_size:
            cluster[n_nodes + t] = cluster.pop(left) + cluster.pop(right)
    clusters = list(cluster.values())
    if sorted_clusters:
        clusters = sorted(clusters, key=len, reverse=True)
    for label, index in enumerate(clusters):
        labels[index] = label
    return labels`

It will be included in the next release.
